app/api/mail.py

```diff
@@ -0,0 +1,128 @@
+    | 500 - something goes wrong
+    :rtype: json
+    """
+
+    response_dictionary = {'status_code': 200}
+    response = {}
+    content = request.get_json(force=True)
+    mail_from = required_param(content, "from", str)
+    mail_to = required_param(content, "to", str)
+    mail_cc = optional_param(content, "cc", str)
+    mail_bcc = optional_param(content, "bcc", str)
+    mail_subj = bleach.clean(optional_param(content, "subject", str))
+    mail_body = bleach.clean(optional_param(content, "text", str))
+    mailto = [mail_to]
+    if mail_cc is not None:
+        mailto += [mail_cc]
+    else:
+        mail_cc = ""
+    if mail_bcc is not None:
+        mailto += [mail_bcc]
+    else:
+        mail_bcc = ""
+
+    # Prepare message
+    message = MIMEMultipart('alternative')
+    message["Subject"] = mail_subj
+    message["From"] = mail_from
+    message["To"] = mail_to
+    message["Cc"] = mail_cc
+    message["Bcc"] = mail_bcc
+    message.attach(MIMEText(mail_body, 'html'))
+
+    try:
+        mail = Email(mail_from=mail_from,
+                     mail_to=mail_to,
+                     mail_cc=mail_cc,
+                     mail_bcc=mail_bcc,
+                     mail_subj=mail_subj,
+                     body=message.as_string(),
+                     status=False,
+                     counter=0
+                     )
+        db.session.add(mail)
+        db.session.commit()
+    except (Exception, exc) as e:
+        db.session().rollback()
+        response_dictionary['status_code'] = 500
+        response_dictionary['error_message'] = "db error"
+        app.logger.error(str(e))
+        return jsonify(**response_dictionary)
+
+    # trying to send email immediately
+    try:
+        response = mailer.send_message(mail_from, mailto, message.as_string())
+        app.logger.info('response')
+        app.logger.info(response)
+        # send_message will return an empty dictionary on success
+        # otherwise we will pass error message to the api
+        if len(response) > 0:
+            response_dictionary['status_code'] = 500
+            response_dictionary['error_message'] = response
+    except Exception as e:
+        response_dictionary['status_code'] = 500
+        response_dictionary['error_message'] = str(e)
+
+    # logging mail delivery status
+    try:
+        mail.status = False if response_dictionary['status_code'] != 200 else True
+        mail.counter = 1
+        response_dictionary['email_id'] = mail.id
+        mail_status = EmailStatus(email_id=mail.id,
+                                  status=json.dumps(response),
+                                  delivered=False if response_dictionary['status_code'] != 200 else True
+                                  )
+        db.session.add(mail_status)
+        db.session.commit()
+    except (Exception, exc) as e:
+        db.session().rollback()
+        response_dictionary['status_code'] = 500
+        response_dictionary['error_message'] = "db error"
```

Comment:
there are 2 cases when you return "db error" in the 1st one no message was sent, in the 2nd one the message was sent, maybe even successfully - you just couldn't write the status to db for some reason. how would the customer know what has happened?

Reply:
fixed. now there is only one possible "db error"